File: 백준/Gold/9007. 카누 선수/카누 선수.py
import sys

input = sys.stdin.readline

INF = float("inf")


tc = int(input())
for _ in range(tc):
    k, n = map(int, input().split())
    weights = [list(map(int, input().split())) for _ in range(4)]
    group1 = []
    group2 = []
    for i in range(n):
        for j in range(n):
            group1.append(weights[0][i] + weights[1][j])
            group2.append(weights[2][i] + weights[3][j])

    group1.sort()
    group2.sort()

    start = 0
    end = n**2 - 1
    answer = INF
    while start < n**2 and 0 <= end:
        s = group1[start] + group2[end]
        if s == k:
            answer = s
            break
        elif abs(s - k) < abs(answer - k):
            answer = s
        elif abs(s - k) == abs(answer - k):
            answer = min(answer, s)

        if s < k:
            start += 1
        else:
            end -= 1

    print(answer)

File: 백준/Gold/9007. 카누 선수/카누 선수.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {

  static StringBuilder sb = new StringBuilder();
  static int testcases;
  static int k, n;
  static int[][] weights;
  static int[] weights1, weights2;

  public static void main(String[] args) throws IOException {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    StringTokenizer st;

    testcases = Integer.parseInt(br.readLine());
    for (int tc = 0; tc < testcases; tc++) {
      st = new StringTokenizer(br.readLine());
      k = Integer.parseInt(st.nextToken());
      n = Integer.parseInt(st.nextToken());
      weights = new int[4][n];
      for (int i = 0; i < 4; i++) {
        st = new StringTokenizer(br.readLine());
        for (int j = 0; j < n; j++) {
          weights[i][j] = Integer.parseInt(st.nextToken());
        }
      }

      weights1 = new int[n * n];
      weights2 = new int[n * n];
      for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
          weights1[n * i + j] = weights[0][i] + weights[1][j];
          weights2[n * i + j] = weights[2][i] + weights[3][j];
        }
      }

      Arrays.sort(weights1);
      Arrays.sort(weights2);

      sb.append(solve()).append("\n");
    }

    output();
  }

  private static int solve() {
    int left = 0;
    int right = n * n - 1;

    int diff = Integer.MAX_VALUE;
    int answer = -1;
    while (left < weights1.length && 0 <= right) {
      int sum = weights1[left] + weights2[right];
      if (sum == k) {
        answer = sum;
        break;
      } else if (sum < k) {
        left++;
      } else {
        right--;
      }

      if (Math.abs(k - sum) < diff) {
        diff = Math.abs(k - sum);
        answer = sum;
      } else if (Math.abs(k - sum) == diff && sum < answer) {
        answer = sum;
      }

    }

    return answer;
  }

  private static void output() throws IOException {
    BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    bw.write(sb.toString());
    bw.flush();
    bw.close();
  }

}